tripal/t4d8
# Group 1: tripal vocab & tripal term

# Metadata
 - Group 1 
 - Issue: #9 
 - Group Issue #1 
 - Includes automated testing? YES
 - Ready to be merged? YES

## Files this PR upgrades
 - TripalVocab
    - [x] TripalVocab.inc
    - [n/a] TripalVocabController.inc (no longer needed)
    - [x] TripalVocabViewsController.inc
 - TripalTerm
    - [ ] tripal.terms.api.inc
    - [x] TripalTerm.inc
    - [n/a] TripalTermController.inc (no longer needed)
    - [x] TripalTermViewsController.inc

# Status Description
This PR makes TripalVocab and TripalTerm functional by upgrading full CRUD operations for both. It also upgrades associated CRUD forms and vocab/term listings. Functional testing is included for both TripalVocab and TripalTerm. I believe TripalTerm/Vocab controllers are no longer needed.

# Testing

## Manual
 - Structure menu should include `Tripal Controlled Vocabularies` and Tripal Controlled Vocabulary Terms` links which link to the Vocab/Term listings.
 - Vocab/Term listings provide links for CRUD operations: Test add/edit/delete
 - Create a view for both terms/vocabularies

## PHPUnit
1. Follow [this guide](https://www.drupal.org/docs/8/phpunit/running-phpunit-tests) to get PHPUnit set-up/configured for use with Drupal 8
2. From your *Drupal Root* run `./vendor/bin/phpunit modules/t4d8/tripal/tests/`

### TripalVocabEntityTest.php:
#### testTripalVocabEntity()
1. Go to Vocab listing as unprivileged user and ensure they get access denied.
2. Login as privileged user; check that "Tripal Controlled Vocabularies" is in the structure listing.
3. Click the link and check it goes to the vocab listing. Also ensure users get a message telling them there is no content yet.
4. Goto the add vocabulary form. Check it has the form fields we expect. Fill it out with a valid unique name and submit. Check for the message indicating the vocab is saved.
5. Go to the Vocab listing. Check there is no message indicating no vocabularies. Also check the name, edit and delete links for the new vocabulary are present in the page.
6. Click the edit link, change the vocabulary name and submit. Check for the message indicating the vocab changes were saved.
7. Go to the Vocab listing. Check the new name, edit and delete links for the changed vocabulary are present in the page. Also check that the old name is not present in the page.
8. Click the delete link and check you get the confirmation page. Leave the delete page without submitting and ensure the vocab wasn't deleted on the listing page.
9. Click delete again. This time confirm delete and check the listing to ensure the vocab is no longer there.

#### testPaths()
For Each route defined for TripalVocab crud ensure a user logged in with a specific set of permissions gets the page status code we expect (i.e. access denied if they don't have sufficient permission and 200 if they do). See `providerTestPaths()` for the list of paths tested.

### TripalTermEntityTest.php:
#### testTripalTermEntity()
1. Go to Term listing as unprivileged user and ensure they get access denied.
2. Login as privileged user; check that "Tripal Controlled Vocabulary Terms" is in the structure listing.
3. Click the link and check it goes to the term listing. Also ensure users get a message telling them there is no content yet.
4. Goto the add vocabulary term form. Check it has the form fields we expect. Fill it out with a valid unique name, accession and newly created vocab and submit. Check for the message indicating the vocab term is saved.
5. Go to the Term listing. Check there is no message indicating no terms. Also check the name, accession, vocab, edit and delete links for the new term are present in the page.
6. Click the edit link, change the term name and submit. Check for the message indicating the term changes were saved.
7. Go to the term listing. Check the new name, old accession, old vocab, edit and delete links for the changed vocabulary are present in the page.
8. Click the delete link and check you get the confirmation page. Leave the delete page without submitting and ensure the term wasn't deleted on the listing page.
9. Click delete again. This time confirm delete and check the listing to ensure the vocab is no longer there.

#### testPaths()
For Each route defined for TripalTerm crud ensure a user logged in with a specific set of permissions gets the page status code we expect (i.e. access denied if they don't have sufficient permission and 200 if they do). See `providerTestPaths()` for the list of paths tested.
